File: material_project0412/scripts/matlab/load_data.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   Description: A basic data loading and ground truth plot
%   Last modified: 2023-09-06
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
clc                         % clear command window
clear all                   % clear workspace
close all                   % close all open figures
%% Load the CSV file
filename = '../../controllers/supervisor/data/ground_truth.csv';
file_odometry = '../../controllers/controller/data/odometry.csv';
data = readtable(filename);
data2 = readtable(file_odometry);

% Strip spaces from column names
data.Properties.VariableNames = strtrim(data.Properties.VariableNames);
data2.Properties.VariableNames = strtrim(data2.Properties.VariableNames);

%% Plot the columns
figure;

subplot(2, 2, 1);
plot(data.x, data.y);
title('Position');
xlabel('x [m]');
ylabel('y [m]');
grid on;

subplot(2, 2, 2);
plot(data.time, data.heading);
title('Heading');
xlabel('time [s]');
ylabel('[rad]');
grid on;

subplot(2, 2, 3);
plot(data.time, data.vel_x);
hold on;
plot(data.time, data.vel_y);
title('Velocity');
xlabel('time [s]');
ylabel('[m/s]');
legend({'$\dot{x}$', '$\dot{y}$'}, 'Interpreter', 'latex');
grid on;

subplot(2, 2, 4);
plot(data.time, data.vel_heading);
title('Angular velocity');
xlabel('time [s]');
ylabel('[rad/s]');
grid on;

sgtitle('Pose Ground Truth');

%% Plot for odometry 

figure()
plot(data2.x, data2.y)
title('Pose estimation (with odometry)');
xlabel('x [m]');
ylabel('y [m]');
grid on;

File: material_project0412/scripts/python/load_data.py
#!/usr/bin/env python3

from matplotlib import pyplot as plt
import pandas as pd

def load_file(filename: str) -> pd.DataFrame:
    '''
    Load a csv file into a pandas dataframe and strip spaces from column names

    Parameters
    ----------
    - `filename`: path to the csv file

    Returns
    -------
    - `df`: pandas dataframe
    '''

    # Load the data 
    df = pd.read_csv(filename, sep=',')

    # strip spaces from column names 
    df.rename(columns=lambda x: x.strip(), inplace=True)

    # drop the last column (empty)
    df = df.drop(df.columns[-1], axis=1)

    return df

def plot_pose_groundtruth(df:pd.DataFrame):
    '''
    Plot the pose ground truth
    '''

    fig = plt.figure(figsize=(6,5))
    shape = (2,2)

    ax = plt.subplot2grid(shape, (0,0))
    ax.set_title('Position')
    ax.plot(df.x, df.y)
    l_min = min(*ax.get_xbound(), *ax.get_ybound())
    l_max = max(*ax.get_xbound(), *ax.get_ybound())
    ax.set_xlim(l_min,l_max)
    ax.set_ylim(l_min,l_max)
    ax.set_xlabel('x [m]')
    ax.set_ylabel('y [m]')
    ax.grid()

    ax = plt.subplot2grid(shape, (0,1))
    ax.set_title('Heading')
    ax.plot(df.time, df.heading)
    ax.set_xlabel('time [s]')
    ax.set_ylabel('[rad]')
    ax.grid()
    
    ax = plt.subplot2grid(shape, (1,0))
    ax.set_title('Velocity')
    ax.plot(df.time, df.vel_x, label='$\dot{x}$')
    ax.plot(df.time, df.vel_y, label='$\dot{y}$')
    ax.set_xlabel('time [s]')
    ax.set_ylabel('[m/s]')
    ax.legend()
    ax.grid()

    ax = plt.subplot2grid(shape, (1,1))
    ax.set_title('Angular velocity')
    ax.plot(df.time, df.vel_heading)
    ax.set_xlabel('time [s]')
    ax.set_ylabel('[rad/s]')
    ax.grid()

    plt.tight_layout()


if __name__ == '__main__':

    df = load_file('controllers/supervisor/data/ground_truth.csv')
    print(df.head())
    
    plot_pose_groundtruth(df)

    plt.show()